### DISCUSSION TITLE
Looking for help?

### DISCUSSION THREAD

**Discussion by NorthernScott** [MightyBOBcnc/nixis#11] [General]
Hey there - I am a new-ish python dev and started work on a similar project when I came across yours. I'm looking to create maps for TTRPG use, so Nixis is pretty much bang on for what I'm hoping for. I'd be interested in helping out if you'd like a hand, but I think I might need a bit of a walk through. Mostly, I've never used kdtrees and only dabbled with meshes. A lot of the work that I've done in this space has been using QGIS/GDAL starting from heightmaps.

Btw, thank you for so extensively annotating your code - it's amazingly helpful.

**Comment by MightyBOBcnc** [comment]
Hi.  I could use all the help I can get, haha.  I'm still quite new to programming.  I haven't been able to work on Nixis for about a year but I do want to resume work this year (hopefully within a few months).  (Note: Dev branch is ahead of Main.)

At the moment the kdtree is only being used as part of the image export, not for any of the generation.  Basically for each pixel of the desired exported image (let's say 2000w x 1000h) that pixel is turned into a latitude/longitude 2D coordinate, and then the LL is converted into an XYZ coordinate in 3D space, and that XYZ coordinate is fed into the kdtree as a query to find the 3 nearest vertices on the mesh to that XYZ position.  From that we can get an average (technically a reverse weighted average) of those vertex heights for that pixel in the exported image.  

[Here's a visualization](https://i.imgur.com/YN9ZOH8.png) that can help show the concept; each red dot is the center of where a pixel on the exported image would be.  There may be a better way to get this information than a kdtree but I'm not experienced enough to know what that may be and the tree was easy enough to implement (you don't actually need to know how a kdtree works to use one) (note: though it does consume a fair amount of resources which is admittedly undesirable).  There might be other uses for the kdtree later.  

(Edit: I guess the kdtree is also being used in a test function for optionally displaying a couple points on the surface but that's not required.)

-----

As for meshes, I'm using an 'icosphere' ([subdivided and normalized icosahedron](https://i.imgur.com/6IDBNhr.png)) because the spacing between vertices is somewhat uniform (moreso than a UV sphere or sphere-ized cube).  In short, the meshzoo library outputs an array of vertices (points) and an array of triangles (cells).  Each vertex on a mesh has its own ID, but the vertices array is **ordered** so the ID is actually the position of the vertex in the array, and the contents of the array is the XYZ coordinate of each vertex, e.g.:
```
vertex 0:  [[ X coord -0.5257311  Y coord 0.85065081  Z coord 0.       ]
vertex 1:   [ X coord 0.52573111  Y coord 0.85065081  Z coord 0.       ]
vertex 2:   [-0.52573111 -0.85065081  0.        ]
vertex 3:   [ 0.52573111 -0.85065081  0.        ]
vertex 4:   [ 0.         -0.52573111  0.85065081]
vertex 5:   [ 0.          0.52573111  0.85065081]
vertex 6:   [ 0.         -0.52573111 -0.85065081]
vertex 7:   [ 0.          0.52573111 -0.85065081]
etc.                                            ]
```

Meshes are [made of triangles](https://user-images.githubusercontent.com/4467749/126717520-d8faf0a1-a502-48a0-bbb9-8f0c986c2e77.png) so the triangle array contains the ID of each vertex that makes up a given triangle, and like the vertex array this is also ordered so the ID of each triangle is its position in the array:
```
triangle 0:  [[ vertex 0  vertex 18 vertex 15]
triangle 1:   [ vertex 18 vertex 11 vertex 37]
triangle 2:   [18 37 15]
triangle 3:   [15 37  5]
triangle 4:   [ 0 15 26]
triangle 5:   [15  5 31]
triangle 6:   [15 31 26]
triangle 7:   [26 31  1]
etc.                   ]
```

Like the kdtree, the triangle array is not actually being used as part of the generation at this time (it is only needed so pyvista can visualize the mesh at the end, or if you want to export a 3D model of the planet) however the triangle array is essentially the foundation for [a graph](https://en.wikipedia.org/wiki/Graph_theory) which will almost certainly be useful later on for other features like calculating water sheds or perhaps some sort of Voronoi diagram or Delaunay triangulation (recommend Red Blob Games blog for procedural generation reading material; I believe he does all his work in javascript, though). 

-----

All of the 'magic' actually happens on the vertices at the moment.  I make a height (elevation) array that you could say is comparable to a heightmap except it is a 1D array, ordered, where each height corresponds to the vertex with the same position in the vertex array:
```
height of vertex 0: [ 8129.28755942  
height of vertex 1:   2371.87542571  
height of vertex 2:   8850.
height of vertex 3:   2314.95553435
height of vertex 4:   -1147.55666521  
height of vertex 5:   -751.24531178
height of vertex 6:   -727.87014505
height of vertex 7:   -1702.32556644
  etc.                            ]
```

The heights are initially all 0.0 and then an opensimplex noise is sampled at multiple octaves to give the terrain its heights (also there's a bit of reshaping of heights related to sea level).  From there I want to do hydraulic erosion but that's where I've hit a roadblock that I've not yet surmounted.  

I'd like to try what's known as a [virtual pipes](https://leatherbee.org/index.php/2018/12/10/climate-2-rivers-and-erosion/) model (side note: there are other possible methods, including graph stuff and particles) but pretty much all of the existing literature and code that other people have written for virtual pipes *expects* that you're using a heightmap, aka a 2D grid of uniform spacing.  As mentioned above, an icosphere is neither a grid nor are the vertices evenly spaced.  Each vertex has 5 or 6 neighbors in basically any direction instead of 4 neighbors in a top, bottom, left, right configuration (note: sometimes virtual pipes also includes the diagonals).  

So what this means is that the model has to be translated into one that is scale and distance and number-of-neighbors agnostic and, well I never took calculus so reading the notation is kinda hard for me, and on top of that I'm still pretty new to programming so it's a bit of a double whammy of difficulty.  It's totally doable, other projects like SotE and Orbis Multiplex have done it, it's just hard for me in particular.  In erosion.py I've tried a few ways but there's something not right with all of them.  The code is so extensively notated precisely because I need to explain it to myself, hah.

**Comment by NorthernScott** [comment]
So, I've been coming at this from the perspective of mapping for an RPG campaign, and I started with 2D heightmaps and manipulating them using GDAL/QGIS. If the doing the erosion and climate modelling on a spherical mesh is causing such trouble, why not do it using a 2D heightmap with a baked in coordinate reference system and only wrap the coordinates to a sphere at the end?

**Reply by MightyBOBcnc** [reply]
**Short answer:** I chose a sphere mesh because I didn't think heightmaps would work and the math for compensating for polar area distortion was intimidating.  (Yes, I'm aware of the irony of choosing a math-heavy project when I'm not good at math.)  

**Longer answer:** Well after my experience with all of the existing heightmap erosion tools not taking spherical distortion into account (https://i.imgur.com/Sn7Tb96.jpg  https://i.imgur.com/pYMKK9k.jpg) I assumed that heightmaps would not be a suitable 'substrate' to work on.  Intuitively it seemed that doing it "for real" on the actual desired spherical shape without any distortion to worry about would be simpler or more elegant because there would be no interpolation math between heightmap pixels or coordinate reference systems except for the final image export step.  i.e Why compensate for distortion if you can do it without distortion in the first place?  

I haven't reevaluated since the original decision.  It might not be the best approach.  

-----

I should also maybe clarify that the current erosion attempts in `erosion.py` aren't actually doing the computations in 3D, which is why I referred to it as "comparable to a heightmap" as each vertex functionally only knows its own height and the height of (and distance to) its 5 or 6 connected neighbors.  It doesn't actually matter where in 3D space a vertex is for those computations.  

Also they're not an implementation of virtual pipes; they're a simpler experiment.  I hadn't gotten to trying virtual pipes yet.  Virtual pipes on a heightmap with a baked in coordinate reference system would also need modification for the high latitude distortion which would be a different type of modification (maybe an easier one) from a modification for a mesh (as IIRC there's a 'look back in time' step in virtual pipes where 3D coords might actually matter on a mesh).

**Comment by NorthernScott** [comment]
I've been acquainting myself with the code - as I said, I'm a beginner - and I noticed that you can probably dump the dependency on Meshzoo (which technically has a 50-euro / yr licence fee even for personal use) and use the Icosphere() function built-in to PyVista. I also think that PyVista includes built in image export functions, which might allow you to avoid using the kd-tree.

I'm currently crawling through the elevation sampling in your custom OpenSimplex code, but the math is a bit  beyond me. I'm thinking for climate and erosion that some of the functions in the GDAL libraries might be helpful.

**Reply by MightyBOBcnc** [reply]
- The version of Meshzoo used by nixis is GPL code that predates the author's attempt to paywall it* so I'm not too bothered by keeping it but that does mean we're locked at version <=0.9.4 (not that there are likely to be any killer new features for icospheres), so investigating PyVista's icosphere isn't a bad idea.

*(and is still available on [conda-forge](https://anaconda.org/conda-forge/meshzoo), or a few [forks on github](https://github.com/MightyBOBcnc/meshzoo), although not pypi anymore.  I updated the requirements.yaml to drop pypi and get it from conda-forge.)

- I think the image export functions in PyVista are only for grabbing camera views or snapshots of the viewport but I would be happy to be proven wrong if it can extract an equirectangular texture from a mesh (in video games and VFX this would be a form of "baking").

- The math in OpenSimplex is beyond me as well, but I don't need to know how the math works to use the library.  😂  I only made very small adjustments to [a fork by ktritz](https://github.com/ktritz/opensimplex/blob/master/opensimplex/opensimplex.py) so that Numba wouldn't error on the class when trying to use it*.  You can actually see discussion about this on [this issue here](https://github.com/lmas/opensimplex/issues/4).  I've been meaning to try out a newer versions of OpenSimplex after the author updated their project but I haven't had the time yet.  See: #10 

*(basically I discarded the use of the class; you could probably diff his and mine to see how I barely did anything to it.  I'm aware you can `@jit` decorate a class with Numba so it won't barf but I haven't gone that deep into the weeds yet.)

As for the way that the elevation sampling is used, there's a lovely guy on YouTube named Sebastian Lague who [describes the concept well](https://www.youtube.com/watch?v=uY9PAcNMu8s&list=PLFt_AvWsXl0cONs3T0By4puYy6GM22ko8&index=3&pp=iAQB).  He's using Unity and C# but the method is the same; sample a coherent noise at multiple octaves to get elevations that [vaguely look like](https://en.wikipedia.org/wiki/Fractal_landscape) the [self-similarity](https://en.wikipedia.org/wiki/Self-similarity) of real terrain at different scales (mountains in particular are rather fractal-like).  See [sample_octaves in terrain.py](https://github.com/MightyBOBcnc/nixis/blob/dev/terrain.py#L32).  Then I do a bit of rescaling and power adjustment (see "Redistribution" [here](https://www.redblobgames.com/maps/terrain-from-noise/#elevation-redistribution)) to further tweak things.  In [dev](https://github.com/MightyBOBcnc/nixis/blob/dev/nixis.py#L337) this all takes place between line 337 and 400 of nixis.py as of the current commit.

This serves as the starting input for erosion, which will improve the look further... once it works...  Eventually, though, some sort of mantle/tectonic uplift sim would be a nice thing to replace the simplex noise as the starting input to the erosion (although the simplex might serve as the input to the mantle OR might still need to be overlaid on top so it might never go away entirely, who knows). 

- I've heard that GDAL is a royal pain in the rear to install.  I think it was one of the reasons I couldn't get [WorldEngine](https://github.com/Mindwerks/worldengine) working the last time I tried (it was dependency hell).  That being said, that was years ago and it may have improved since then; probably worth investigating.

**Reply by NorthernScott** [reply]
Hey, thanks for the reply.

I've got a version working using the pyvista icosphere - it was a piece of cake once I started to understand the basics of pyvista. I've also implemented the latest OpenSimplex module. I've had to remove the numba @njit dependency (though I've kept prange) due to the type error you mentioned. That said, since Numba is now baked into the osi module, it's still running pretty fast. I have yet to incorporate the rescaling but that's what I'm working on next. And thanks for pointing me to the terrain gen resources. I've encountered Sebastian's work but didn't realize that he had a C# tutorial on this. I think I may have actually found it a couple of years ago but it would have been beyond me at that point.

And yes, I've worked with GDAL a bit and it is a huge pita to install. That said, it has some incredible functionality that would be super helpful. I've used it with GeoTIFFs in projected coordinates to do hydrological modelling (in QGIS).

Agree with you also on the dependency hell of WorldEngine. That said, I actually prefer the terrain generation that you're getting without the plate tectonics. WorldEngine's tectonics never felt fully baked to me. I have yet to find a good Python implementation of plate tectonics.

I'd be interested in assisting if I can. Do you want help, and if so, where?

**Reply by MightyBOBcnc** [reply]
Totally agreed on the tectonics in WorldEngine.  I think the rifting events and sudden direction changes felt artificially random, not like they were being driven by an underlying cause like mantle plumes or crust stress that would lead to organic looking shapes and gradual change over time instead of instant direction changes.  They also happened for the entire map at once instead of only one region at a time where something like a mantle hot spot changed.  (It should be noted that instead of actually moving the plates like WorldEngine an alternate method is to fake tectonics [with a Voronoi diagram](https://www.redblobgames.com/x/1843-planet-generation/) split into regions and then you fudge the plate interactions like subduction at the boundaries.)

**TL;DR:** If you want to help, the world is kind of your oyster at the moment.  There is much to do and much to choose from.  I think it might be down to which parts interest you the most.  

* I've tried erosion a few ways (erosion.py), some of them seemed to have promise, but something is wrong with my math.  I'd feel guilty to saddle you with that mess. 
* In climate.py I started the process of accumulating incoming solar insolation in the lower half of the file and can make nice output images and snapshots like so (note: image scale is relative here, with min being black and max being white, otherwise the instant "moment in time" would be essentially solid black compared to annual): 

One moment in time (with axial tilt):
![848_320_instant_insol](https://github.com/MightyBOBcnc/nixis/assets/4467749/6afcb0d0-82f0-4aaa-b74d-0e5ff0ce04be)

One "day" (there are two different ways to get this, one is a function that brute force adds X number of "moments" with rotation, the other function involves interpolation of a narrow slice across the whole planet):
![848_320_daily_insolation](https://github.com/MightyBOBcnc/nixis/assets/4467749/97b91266-40f4-4e0b-bb6b-e55da76d8ae4)

Full "year" (Y of the "days" added together, incorporating change due to axial tilt): 
![848_320_annual_insolation](https://github.com/MightyBOBcnc/nixis/assets/4467749/248d6c1f-406e-48c4-b04a-f0718f6e787a) 

At the moment these are sort of like strength masks, not containing actual insolation wattage, but if we multiply their values times an actual Total Solar Irradiance (TSI) constant (which for Earth is about 1370 watts per square meter, at the top of the atmosphere) it should give us the right value for any point on the planet at any time, or averaged over time.. almost.. I think it's off by ~20 watts(?), but it's close.   

* What hasn't been done yet is take that energy value at the top of the atmosphere and convert it into temperature at the planet surface.  In the top half of climate.py there are some experiments with trying to find the equilibrium temperature for a block of a material (such as water) but again there's some "fun" math to be done and some problems with the current functions when trying to do this with large spans of time, so that's an area you could consider.  The intent is to get the heat capacity and equilibrium temperature for various material types (water, snow, stone, etc.) as one of the things that informs the surface temperature, along with what will probably be a fairly standard 2-Layer Atmosphere sim, which I hope will strike a balance between simplicity and accuracy.  [Resource](https://leatherbee.org/index.php/2018/12/17/climate-3-insolation-and-black-body-radiation/) ([Broken embed from resource](https://leatherbee.org/wp-content/uploads/2018/12/radiationWithVariation.gif)) [More resource](https://personal.ems.psu.edu/~dmb53/DaveSTELLA/climate/climate_modeling_1.htm)
* Then there's the presentation.  I moved the PyVista stuff into gui.py, and it needs work as well.  The process for visualizing the scalar color values for [elevation](https://i.imgur.com/zAqXAFS.png), [insolation](https://i.imgur.com/HOoFnda.png), etc. is very very manual at the moment, which could be improved upon.  Eventually it might be best to have buttons for switching between scalars, or swapping out exported textures if PyVista supports that.  Right now the scalars are used as vertex color, I believe, and I'm not sure if PyVista can swap vertex color on the fly.
* There's the question of image export and how to do it.  And a bug with the current exporter for large images that I haven't looked into yet, see: #4.  There's also an annoying missing feature in the `pillow` image library where it doesn't support 16-bit RGB png images (only grayscale, which is okay for elevation heightmaps but will be a problem if we want higher bit-depth RGB images later).  `opencv` has that support but it's a big library and I'm kind of loathe to install a huge library to only use one small feature (bit of a hypocrite, though, as Nixis uses matplotlib for like 1 color function; I'd actually like to kill matplotlib as a dependency for that reason, if possible).  There might be other candidate png image libraries.  (There's also the question of additional image format support, such as geotiffs.  Pillow supports a variety of formats including tiff but not all image formats are suitable for every purpose for bit depth reasons and feature support and such.)
* There are ToDos all over pretty much every file, which include fixes, replacement of 'hacks,' enhancements, entire features, and more.

Do you have your version on git?

**Reply by NorthernScott** [reply]
Cool. I am going to dive in and see what I can do. Since I am a beginner, I was thinking that I might work on the CLI and GUI, bringing in logging, tests, etc. Quality of life stuff that won't gibble your love work. ;)

A couple of thoughts:

- I can look into the insolation stuff and I'll go over the resources you shared on atmosphere sim.
- As for insolation / albedo of different surfaces, how would you feel about using two average values, one for land and one for water? I'm just thinking that to determine whether a surface is bare rock, grass, trees, snow, etc., we would first need to understand the biome at play which we can only do once we have baseline temperature and precipation values for each cell. Maybe we can eventually build a sim that does multiple passes, but I'd vote for starting with a simpler model.
- As for 16-bit outputs, I've definitely noticed some oddity in that department. That said, we could use 32-bit image exports.
- In terms of image size, I've been looking at a resolution of 1px / km, so 30,000 x 15,000 is probably the lowest pixel size I'd want to use for a roughly 1px to 1km resolution. 36,000 x 18,000 is evenly divisible by 360 and multiples thereof, but is approximaetly 130% larger than necessary. 40,000 x 20,000 is closest to the actual circumferential dimensions of the earth but because Earth is not a perfect sphere, yields approximately 300 million more pixels than required to model the surface area (~160% larger than necessary).

**Reply by MightyBOBcnc** [reply]
Ah yes I totally forgot to mention logging and unit tests.  I'm sure you've noticed that the code is loaded to the gills with print statements. Logging is [one of the ToDos](https://github.com/MightyBOBcnc/nixis/blob/fb49db0c7015a6bf9c6e4496901754bf4e0e1efa/nixis.py#L69) in that huge list of ToDos in the main file.  (I've also flirted with the idea of [adding tqdm](https://github.com/tqdm/tqdm) for progress bars on certain functions but that's definitely a frilly optional thing, not a necessity, plus I'm not sure of its Numba compatibility or if it has performance impacts.)  Splitting out all of the argparse stuff and initial setup between `# Initial setup stuff` and `# Start the party` is probably a good idea.

* Yes, as a first step I completely agree with implementing only a generic land and water for the initial surfaces.  Other surface types can be bolted on later.
* I have only a little experience with 32-bit (per channel) images used in computer graphics (I have more with 16-bit used in computer graphics) but I'm not opposed to them.  It's mostly about compatibility.  I envision 2 "types" of image exports from Nixis that we could call "data" and "pretty."  Data images would be those that we export at a higher bit depth that preserves more fine-grained information that can be useful in other programs, while Pretty images are simply those exported at your typical 8-bits-per-channel common to most images shared online.  You could export the same image in either type.  In other words Data images are functional and have utility, such as an elevation map with a 65000+ levels of height that you could import into some 3D tool, and Pretty images are for display, such as an elevation map with 256 levels of height that you share with your D&D group (and which also might be in a lossy format like jpg) where super precision isn't needed.
* Yes, roughly 1 px/km is my target detail size.  If Nixis could go higher that would be icing on top, but not required.  Given the huge image dimensions it might be prudent to have the ability to export as 1 image or divided into 8 chunks like the [Visible Earth](https://visibleearth.nasa.gov/images/73934/topography).  I'm okay with modeling planets as spheres even though they're technically oblate spheroids. 

I see your PR.  Thanks, I'll have a look.
